lib/lookup/src/lookup_buf/mod.rs

```diff
@@ -0,0 +1,399 @@
+        self.segments.get(index)
+    }
+
+    /// Push onto the internal list of segments.
+    #[instrument(level = "trace", skip(segment))]
+    pub fn push_back(&mut self, segment: impl Into<SegmentBuf>) {
+        self.segments.push_back(segment.into());
+    }
+
+    #[instrument(level = "trace")]
+    pub fn pop_back(&mut self) -> Option<SegmentBuf> {
+        self.segments.pop_back()
+    }
+
+    #[instrument(level = "trace", skip(segment))]
+    pub fn push_front(&mut self, segment: impl Into<SegmentBuf>) {
+        self.segments.push_front(segment.into())
+    }
+
+    #[instrument(level = "trace")]
+    pub fn pop_front(&mut self) -> Option<SegmentBuf> {
+        self.segments.pop_front()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn iter(&self) -> std::collections::vec_deque::Iter<'_, SegmentBuf> {
+        self.segments.iter()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn len(&self) -> usize {
+        self.segments.len()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    #[instrument(level = "trace")]
+    pub fn is_root(&self) -> bool {
+        self.is_empty()
+    }
+
+    /// Raise any errors that might stem from the lookup being invalid.
+    #[instrument(level = "trace")]
+    pub fn is_valid(&self) -> Result<(), LookupError> {
+        Ok(())
+    }
+
+    #[instrument(level = "trace")]
+    pub fn clone_lookup(&self) -> Lookup {
+        Lookup::from(self)
+    }
+
+    #[instrument(level = "trace")]
+    pub fn from_str(value: &str) -> Result<LookupBuf, LookupError> {
+        Lookup::from_str(value).map(|l| l.into_buf())
+    }
+
+    pub fn from_segments(segments: Vec<SegmentBuf>) -> Self {
+        Self {
+            segments: segments.into_iter().collect(),
+        }
+    }
+
+    /// Return a borrow of the SegmentBuf set.
+    #[instrument(level = "trace")]
+    pub fn as_segments(&self) -> &VecDeque<SegmentBuf> {
+        &self.segments
+    }
+
+    /// Return the SegmentBuf set.
+    #[instrument(level = "trace")]
+    pub fn into_segments(self) -> VecDeque<SegmentBuf> {
+        self.segments
+    }
+
+    /// Merge a lookup.
+    #[instrument(level = "trace")]
+    pub fn extend(&mut self, other: Self) {
+        self.segments.extend(other.segments)
+    }
+
+    /// Returns `true` if `needle` is a prefix of the lookup.
+    #[instrument(level = "trace")]
+    pub fn starts_with(&self, needle: &LookupBuf) -> bool {
+        needle.iter().zip(&self.segments).all(|(n, s)| n == s)
+    }
+
+    /// Create the possible fields that can be followed by this lookup.
+    /// Because of coalesced paths there can be a number of different combinations.
+    /// There is the potential for this function to create a vast number of different
+    /// combinations if there are multiple coalesced segments in a path.
+    #[instrument(level = "trace")]
+    pub fn to_alternative_components(&self) -> Vec<Vec<&str>> {
```

Comment:
From what I could find, this function is only used to do one of two things:

- Get the first path segment
- Get the first two path segments

It seems that we could earn back a few allocations if we were to drop this method and have one or more specifically tailored to the above use-cases?

Reply:
Have added a limit parameter.